fix: send webhook reports when altstorage is used

vmax opened the pull request:
## Description

## Related issues
- Close #2682

Remove this section if you don't have related PRs.

## Checklist
- [X] I've read the [guidelines for contributing](https://github.com/aquasecurity/trivy-operator/blob/main/CONTRIBUTING.md) to this repository.
- [x] I've added tests that prove my fix is effective or that my feature works.
- [X] I've updated the [documentation](https://github.com/aquasecurity/trivy-operator/tree/main/docs) with the relevant information (if needed).
- [X] I've added usage information (if the PR introduces new options)
- [X] I've included a "before" and "after" example to the description (if the PR is a user interface change).

vmax committed:
fix: send webhook reports when altstorage is used

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/aquasecurity/trivy-operator?pullRequest=2683) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/aquasecurity/trivy-operator?pullRequest=2683) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/aquasecurity/trivy-operator?pullRequest=2683) it.</sub>

simar7 commented on pkg/compliance/clustercompliancereport.go:
```diff
@@ -86,6 +87,13 @@ func (r *ClusterComplianceReportReconciler) generateComplianceReport(ctx context
 				}
 				log.Info("Cluster compliance report written", "path", reportPath)
 
+				// Send webhook notification for compliance report
+				if r.Config.WebhookBroadcastURL != "" {
+					if err := webhook.SendWebhookReport(&report, r.Config); err != nil {
+						log.Error(err, "Failed to send compliance report via webhook")
+					}
+				}
+
```
Thanks for the PR! Can you add a small unit test that would exercise this? One idea to do so would be to create an http test server and then pass its URL to this function. That way we can assert the results within the test server.

vmax replied:
@simar7 added a test in https://github.com/aquasecurity/trivy-operator/pull/2683/commits/3ba57d23dde20f542cc480b386d1b7cb0f8b098e; plz review

vmax committed:
fix linter errors

vmax committed:
webhook test

vmax committed:
address lint comments

vmax committed:
address cyclomatic complexity

vmax committed:
send complete reports

vmax committed:
send same data as written to the file

afdesk commented on pkg/webhook/webhookreporter.go:
```diff
@@ -103,6 +118,31 @@ func (r *WebhookReconciler) reconcileReport(reportType client.Object) reconcile.
+	req.Header = headerValues
+
+	resp, err := hc.Do(req)
+	defer func() {
+		if resp != nil {
+			_ = resp.Body.Close()
+		}
+	}()
+	if err != nil {
+		return fmt.Errorf("failed to send reports to endpoint: %w", err)
+	}
```
I'm unsure the `if resp != nil` is needed, IIRC if there is no error, `resp` can't be nil. need to check it

afdesk replied:
Also I think we need to log `status` and `message` from a response.
as for me, we shouldn't skip this information 

@simar7 WDYT?

vmax replied:
Applied, thanks — the response body is now closed via defer once the request succeeds.

vmax replied:
You're right — when there's no error `resp` can't be nil. Reordered to check the error first, then `defer resp.Body.Close()` without the nil guard.

vmax replied:
Agreed. Non-2xx responses now return an error containing the status line and response body, so the caller logs it instead of silently dropping that information.

afdesk commented on pkg/webhook/webhookreporter.go:
```diff
@@ -68,9 +68,24 @@ func (r *WebhookReconciler) SetupWithManager(mgr ctrl.Manager) error {
 
+// SendWebhookReport sends a report directly via webhook without going through CRD reconciliation
+// This is used when AltReportStorageEnabled is true and reports are written to filesystem
+func SendWebhookReport(reportData []byte, config etc.Config, log logr.Logger) {
```
I'd improve a name for this function.
something like `SendWebhookReportIfAvailable`, in favor of other functions in `config`: https://github.com/aquasecurity/trivy-operator/blob/main/pkg/plugins/trivy/config.go#L375

@simar7 WDYT?

vmax replied:
Good call — renamed to `SendWebhookReportIfAvailable` (it's a no-op when no webhook URL is configured) and updated all call sites.

afdesk reviewed (commented):
@vmax thanks for your cool PR.
I left a few small comments for discussion

kovasaurus commented:
bump, any progress with this?

vmax committed:
address review feedback and fix rebase conflicts

- rename SendWebhookReport -> SendWebhookReportIfAvailable (afdesk)
- check response status and surface body on non-2xx webhook replies (afdesk)
- close response body only when the request succeeds (resp is non-nil)
- keep upstream flow: still generate the compliance report after alt-storage write
- have streamReportToFile return the encoded bytes so the webhook sends the
  exact payload written to disk (fixes undefined reportData)

vmax commented:
Rebased onto latest `main` and addressed @afdesk's feedback:

- Renamed `SendWebhookReport` → `SendWebhookReportIfAvailable` (no-op when no webhook URL is set).
- Webhook now checks the response status and returns an error with the status line + body on non-2xx replies instead of discarding it.
- Response body is closed only after a successful request (reordered, since `resp` is non-nil when there's no error).

Ready for another look 🙏
